kernel/fatal.c

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright (c) 2019 Intel Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel.h>
+#include <kernel_internal.h>
+#include <sys/printk.h>
+#include <sys/__assert.h>
+#include <arch/cpu.h>
+#include <logging/log_ctrl.h>
+#include <logging/log.h>
+#include <fatal.h>
+
+/* LCOV_EXCL_START */
+FUNC_NORETURN __weak void z_arch_system_halt(unsigned int reason)
+{
+	ARG_UNUSED(reason);
+
+	/* TODO: What's the best way to totally halt the system if SMP
+	 * is enabled?
+	 */
+
+	(void)z_arch_irq_lock();
+	for (;;) {
+		k_cpu_idle();
+	}
+}
+/* LCOV_EXCL_STOP */
+
+/* LCOV_EXCL_START */
+__weak void k_sys_fatal_error_handler(unsigned int reason,
+				      const z_arch_esf_t *esf)
+{
+	ARG_UNUSED(esf);
+
+	LOG_PANIC();
+	z_fatal_print("Halting system");
+	z_arch_system_halt(reason);
+	CODE_UNREACHABLE;
+}
+/* LCOV_EXCL_STOP */
+
+#if defined(CONFIG_LOG) || defined(CONFIG_PRINTK)
+static const char *thread_name_get(struct k_thread *thread)
+{
+	const char *thread_name = k_thread_name_get(thread);
+
+	if (thread_name == NULL || thread_name[0] == '\0') {
+		thread_name = "unknown";
+	}
+
+	return thread_name;
+}
+
+static const char *reason_to_str(unsigned int reason)
+{
+	switch (reason) {
+	case K_ERR_CPU_EXCEPTION:
+		return "CPU exception";
+	case K_ERR_SPURIOUS_IRQ:
+		return "Unhandled interrupt";
+	case K_ERR_STACK_CHK_FAIL:
+		return "Stack overflow";
+	case K_ERR_KERNEL_OOPS:
+		return "Kernel oops";
+	case K_ERR_KERNEL_PANIC:
+		return "Kernel panic";
+	default:
+		return "Unknown error";
+	}
+}
+
+void z_fatal_print(const char *fmt, ...)
+{
+	va_list ap;
+
+	LOG_PANIC();
+
+	va_start(ap, fmt);
+	if (IS_ENABLED(CONFIG_LOG)) {
+		struct log_msg_ids src_level = {
+			.level = LOG_LEVEL_ERR
```

Comment:
source is not set, it will be printed with random source prefix. I think it's better to switch to the macro as suggested in the comment above.

Reply:
fixed